### PR TITLE
feat(spanner): add cmek backup support - ext of #8142

### PR DESCRIPTION
**Pull request by jiren**
This is same code base of #8142  created by @hengfengli with added do example.

### CODE REVIEW

**Review by skuruppu — APPROVED** [review]

**Comment by skuruppu** [comment]
@jiren you can rebase this PR now since the code generation PR has been merged.

**Comment by jiren** [comment]
> @jiren you can rebase this PR now since the code generation PR has been merged.

Done.

File: google-cloud-spanner/lib/google/cloud/spanner/database.rb [hunk 1]
```diff
@@ -443,14 +451,38 @@ def database_operations filter: nil, page_size: nil
...
+        #   job.done? #=> true
+        #
+        #   if job.error?
+        #     status = job.error
+        #   else
+        #     backup = job.backup
+        #   end
+        #
+        def create_backup backup_id, expire_time,
+                          version_time: nil, encryption_config: nil
```
Comment (thiagotnunes) [inline google-cloud-spanner/lib/google/cloud/spanner/database.rb:477]: The `encryption_config` of a `backup` has different options than a `database`'s encryption config. See https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/backup.proto#L157-L158 and https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/backup.proto#L425-L460. This means that a backup encryption config can be one of the following:

- [CustomerManagedEncryption](https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/backup.proto#L444): We need to specify a `kms_key_name` when this option is used.
- [GoogleDefaultEncryption](https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/backup.proto#L440): no key should be specified.
- [UseDatabaseEncryption](https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/backup.proto#L437): no key should be specified.

File: google-cloud-spanner/lib/google/cloud/spanner/service.rb [hunk 2]
```diff
@@ -545,12 +547,13 @@ def list_backup_operations instance_id,
 
         def restore_database backup_instance_id, backup_id,
                              database_instance_id, database_id,
-                             call_options: nil
+                             call_options: nil, encryption_config: nil
```
Comment (thiagotnunes) [inline google-cloud-spanner/lib/google/cloud/spanner/service.rb:550]: The restore database `encryption_config` is different than the `backup` and the `create database` encryption configs. See https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/spanner_database_admin.proto#L740-L741 and https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/spanner_database_admin.proto#L744-L778. The encryption here can be one of the following:

- [CustomerManagedEncryption](https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/spanner_database_admin.proto#L761): a key should be specified.
- [GoogleDefaultEncryption](https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/spanner_database_admin.proto#L757): no key should be specified.
- [UseBackupEncryption](https://github.com/googleapis/googleapis/blob/master/google/spanner/admin/database/v1/spanner_database_admin.proto#L754): no key should be specified.

**Review by thiagotnunes — CHANGES_REQUESTED** [review]
There are different encryption configurations, depending if the operation is a `create database`, `create backup` or `restore database`. All of them accept the `CustomerManagedEncryption`, which has a `kms_key_name` specified, but the `backup` and `restore` accept different configurations.

@jiren could you make sure the implementation work with the different configurations for the different operations (see my comments in the code for the support of each operation)?

cc @larkee

File: google-cloud-spanner/lib/google/cloud/spanner/database.rb (hunk 1 above)
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/database.rb:477]: ok, I will update it.

**Comment by larkee** [comment]
@jiren Friendly ping for an update :)

**Comment by jiren** [comment]
> @jiren Friendly ping for an update :)

Partially completed. I will complete changes mostly be tomorrow.

**Comment by larkee** [comment]
> Partially completed. I will complete changes mostly be tomorrow.

Awesome, thanks :+1:

**Comment by jiren** [comment]
> > Partially completed. I will complete changes mostly be tomorrow.
> 
> Awesome, thanks 👍

@larkee  Updated changes.

File: google-cloud-spanner/lib/google/cloud/spanner/backup.rb [hunk 3]
```diff
@@ -251,6 +251,29 @@ def delete
...
+        #   the encryption type and key resources in Cloud KMS used to
+        #   encrypt/decrypt the database to restore to. If this field is not
+        #   specified, the restored database will use the sam encryption
+        #   configuration as the backup by default. Optional. The following
+        #   settings can be provided:
+        #
+        #   * `:kms_key_name` (String) The name of KMS key to use which should
+        #     be the full path, e.g., `projects/<project>/locations/<location>\
+        #     /keyRings/<key_ring>/cryptoKeys/<kms_key_name>`
+        #      This field should be set only when encryptioon type
```
Comment (larkee) [inline google-cloud-spanner/lib/google/cloud/spanner/backup.rb:264]:
```suggestion
        #      This field should be set only when encryption type
```

File: google-cloud-spanner/lib/google/cloud/spanner/backup.rb [hunk 4]
```diff
@@ -251,6 +251,29 @@ def delete
         #   project and have the same instance configuration as the instance
         #   containing the source backup. Optional. Default value is same as a
         #   backup instance.
+        # @param [Hash] encryption_config An encryption configuration describing
+        #   the encryption type and key resources in Cloud KMS used to
+        #   encrypt/decrypt the database to restore to. If this field is not
+        #   specified, the restored database will use the sam encryption
```
Comment (larkee) [inline google-cloud-spanner/lib/google/cloud/spanner/backup.rb:257]:
```suggestion
        #   specified, the restored database will use the same encryption
```

File: google-cloud-spanner/lib/google/cloud/spanner/database.rb [hunk 5]
```diff
@@ -424,6 +424,18 @@ def database_operations filter: nil, page_size: nil
         #   * `:kms_key_name` (String) The name of KMS key to use which should
         #     be the full path, e.g., `projects/<project>/locations/<location>\
         #     /keyRings/<key_ring>/cryptoKeys/<kms_key_name>`
+        #     This field should be set only when encryptioon type
```
Comment (larkee) [inline google-cloud-spanner/lib/google/cloud/spanner/database.rb:427]:
```suggestion
        #     This field should be set only when encryption type
```

**Review by larkee — COMMENTED** [review]
LGTM. Just some typo fix suggestions

File: google-cloud-spanner/lib/google/cloud/spanner/database.rb (hunk 5 above)
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/database.rb]: done

File: google-cloud-spanner/lib/google/cloud/spanner/database.rb [hunk 6]
```diff
@@ -443,14 +463,48 @@ def database_operations filter: nil, page_size: nil
...
+        #   job.reload! # API call
+        #   job.done? #=> true
+        #
+        #   if job.error?
+        #     status = job.error
+        #   else
+        #     backup = job.backup
+        #   end
+        #
+        def create_backup backup_id, expire_time,
```
Comment (thiagotnunes) [inline google-cloud-spanner/lib/google/cloud/spanner/database.rb:491]: What happens if we try to create a database / backup / restore with an invalid `encryption_type`. Something like:

```ruby
spanner = Google::Cloud::Spanner.new
database = spanner.database "my-instance", "my-database"
kms_key_name = "projects/<project>/locations/<location>/keyRings/<key_ring>/cryptoKeys/<kms_key_name>"
encryption_config = { encryption_type: :INVALID_ENCRYPTION }

job = database.create_backup "my-backup", Time.now + 36000, encryption_config: encryption_config
```

Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/database.rb:491]: `encryption_type` is enum  so if value is not present in enum then Protobuf will reject it with error.

`Unknown symbol value for enum field 'encryption_type'.`

Comment (thiagotnunes) [reply google-cloud-spanner/lib/google/cloud/spanner/database.rb:491]: Cool, thanks for the information.

**Comment by thiagotnunes** [comment]
@jiren Will integration tests for CMEK encryption be added when we created the samples (for create database, create backup and restore database)?

**Comment by jiren** [comment]
> @jiren Will integration tests for CMEK encryption be added when we created the samples (for create database, create backup and restore database)?

Integration tests updated for CMEK for create backup and restore. 

In this PR pending is few more checks in integration test  - after restore database check encryption config(this is dependant on #8167 and create database.

I will update create_databse integration tests in #8167

Other pending thing is samples in https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/master/spanner/spanner_samples.rb

**Review by thiagotnunes — APPROVED** [review]

**Comment by larkee** [comment]
`Kokoro - OSx` check is failing `doctest`

**Comment by thiagotnunes** [comment]
I think it was a flake, rebuilding now.

**Comment by jiren** [comment]
> I think it was a flake, rebuilding now.


Mac tests are failing due to known issue specific to the ruby 3.0.0
https://github.com/grpc/grpc/issues/25222

**Comment by dazuma** [comment]
Merging #8167 resulted in a conflict with this one.

**Comment by jiren** [comment]
@dazuma @thiagotnunes Rebased branch with master and removed conflicts.

**Pull request merged**